ocp: Add list acknowledgement gate support

jharrington22 opened the pull request:
Add support for to list OCP gates

vkareh commented on cmd/list/gates/cmd.go:
```diff
@@ -0,0 +1,114 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
2020 -> 2022

vkareh commented on cmd/list/gates/cmd.go:
```diff
@@ -0,0 +1,114 @@
+	}()
+
+	// if args.version == "" {
+	// 	reporter.Errorf("OCP Version required")
```
You can always add `Cmd.MarkFlagRequired("version")` in the `init` function

vkareh commented on pkg/ocm/gates.go:
```diff
@@ -0,0 +1,81 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
```
2020 -> 2022

vkareh commented on pkg/ocm/gates.go:
```diff
@@ -0,0 +1,81 @@
+	versionGatesRequest := c.ocm.ClustersMgmt().V1().VersionGates()
+
+	page := 1
+	size := 100
```
I doubt we'll exceed 100 gates for now, but might be worth checking if the API supports `-1` to return all items

jharrington22 replied:
We're paging through the results here so it shouldn't matter? Or is it a faster? Happy to change it

vkareh commented on pkg/ocm/gates.go:
```diff
@@ -0,0 +1,81 @@
+	size := 100
+
+	for {
+		response, err := versionGatesRequest.List().
+			Page(page).
+			Size(size).
+			Send()
```
Does this API support filtering by version? If so, we should allow this, otherwise we'd need to fetch all gates for all versions then filter on the client.

vkareh commented on cmd/list/gates/cmd.go:
```diff
@@ -0,0 +1,114 @@
+	Short:   "List available OCP Gates",
+	Long:    "List available OCP Gates.",
+	Example: `  # List all available OCP Gates 
+  rosa list gates --version 4.9`,
```
I think it makes sense to allow a `--cluster` flag and show whether the ACKs have been set for its gates.

jharrington22 replied:
Done

vkareh reviewed (commented):
Thanks for starting this @jharrington22!

zgalor commented:
@jharrington22 this only list gates. No ack support, right?

jharrington22 commented:
> @jharrington22 this only list gates. No ack support, right?

No it doesn't yet though I added some local changes I had which would add the ack support. I was wondering if the gates should be its own sub command?

zgalor commented:
> > @jharrington22 this only list gates. No ack support, right?
> 
> No it doesn't yet though I added some local changes I had which would add the ack support. I was wondering if the gates should be its own sub command?

We don't need a sub command for gates. We'll handle them as part of the upgrade command

jharrington22 commented:
List OCP gates:

```
$ rosa list gates --version 4.9 --gate ocp                   
Gate Description                                                                                                                                                                                                                                             STS      OCP Version    Documentation URL
test                                                                                                                                                                                                                                                         false    4.9            
OpenShift removes several Kubernetes APIs in OpenShift 4.9. To help prevent issues with workloads and tools after upgrading, an administrator is required to provide manual acknowledgement before the cluster can be upgraded from OpenShift 4.8 to 4.9.    false    4.9            https://access.redhat.com/articles/6329921
```

List STS gates:

```
$ rosa list gates --version 4.9 --gate sts
Gate Description                                                                                                                                                                                                                                                                                                         STS     OCP Version    Documentation URL
OpenShift STS clusters include new required cloud provider permissions in OpenShift 4.9. To prevent issues with the cluster upgrade, an administrator is required to provide manual acknowledgement that all STS roles are updated with 4.9 permissions before the cluster can be upgraded from OpenShift 4.8 to 4.9.    true    4.9            https://docs.openshift.com/rosa/rosa_getting_started_sts/rosa_creating_a_cluster_with_sts/rosa-sts-about-iam-resources.html
```

List all gates

```
rosa list gates --version 4.9           
Gate Description                                                                                                                                                                                                                                                                                                         STS      OCP Version    Documentation URL
OpenShift STS clusters include new required cloud provider permissions in OpenShift 4.9. To prevent issues with the cluster upgrade, an administrator is required to provide manual acknowledgement that all STS roles are updated with 4.9 permissions before the cluster can be upgraded from OpenShift 4.8 to 4.9.    true     4.9            https://docs.openshift.com/rosa/rosa_getting_started_sts/rosa_creating_a_cluster_with_sts/rosa-sts-about-iam-resources.html
test                                                                                                                                                                                                                                                                                                                     false    4.9            
OpenShift removes several Kubernetes APIs in OpenShift 4.9. To help prevent issues with workloads and tools after upgrading, an administrator is required to provide manual acknowledgement before the cluster can be upgraded from OpenShift 4.8 to 4.9.                                                                false    4.9            https://access.redhat.com/articles/6329921
```

pvasant commented:
@jharrington22  are we planning to include this the upcoming release? I m currently aiming for Jan 25

jharrington22 commented:
@pvasant updated you can now list gates for a cluster providing the `--cluster` arg along with the ability to just list available gates for a given OCP version.

pvasant commented on cmd/list/gates/cmd.go:
```diff
@@ -94,47 +126,94 @@ func run(cmd *cobra.Command, _ []string) {
+			os.Exit(1)
+		}
+
+		if len(versionGates) == 0 {
```
We might need to add this out so it applies for empty list from cluster

pvasant commented:
> @pvasant updated you can now list gates for a cluster providing the `--cluster` arg along with the ability to just list available gates for a given OCP version.
> 
> ```
> rosa list gates -c <cluster_id> --version 4.9.15
> Gate Description                                                                                                                                                                                                                                           STS      OCP Version    Documentation URL
> OpenShift removes several Kubernetes APIs in OpenShift 4.9. To help prevent issues with workloads and tools after upgrading, an administrator is required to provide manual acknowledgement before the cluster can be upgraded from OpenShift 4.8 to 4.9.    false    4.9            https://access.redhat.com/articles/632992
> ```

@jharrington22  I thought it will display the gates acknowledged for the cluster than missing gates. Maybe we need to change it to rosa list availablegates . I m not sure/. @arendej @vkareh WDYT? 

Also the description is so big may be some kind of word wrap might help

jharrington22 commented:
@pvasant I added word wrapping reserving 30% of terminal space for the description of the gates. I added package that uses standard lib's to get the terminal size, supports both Windows/Darwin/Linux. I also changed tabwriter spacing from tabs `\t` to  8 white spaces `' '` so it has better support in different terminals, a change we should make across other commands.  

Looks like the gates have changed in stage at least with gate description lengths

```
$ rosa list gates --gate ocp --version 4.9.15 --region us-east-1
Gate Description                                             STS    OCP Version  Documentation URL                                                                                                               
OpenShift removes several Kubernetes APIs in OpenShift 4.9.  false  4.9          https://access.redhat.com/documentation/en-us/openshift_container_platform/4.9/html/updating_clusters/updating-cluster-prepare  

$ rosa list gates --gate sts --version 4.9.15 --region us-east-1       
Gate Description                                                        STS   OCP Version  Documentation URL  
OpenShift STS clusters include new required cloud provider permissions  true  4.9                             
in OpenShift 4.9.                                                                                             
$ rosa list gates --profile jamesh2 --region us-east-1 --version 4.9.15
Gate Description                                                        STS    OCP Version  Documentation URL                                                                                                               
OpenShift removes several Kubernetes APIs in OpenShift 4.9.             false  4.9          https://access.redhat.com/documentation/en-us/openshift_container_platform/4.9/html/updating_clusters/updating-cluster-prepare  
OpenShift STS clusters include new required cloud provider permissions  true   4.9                                                                                                                                          
in OpenShift 4.9.                                                                                                                                                                                                           
$ rosa list gates -c 1pusm2bqji12pf19023v14s2160q9e69 --profile jamesh2 --region us-east-1 --version 4.9.15
Gate Description                                             STS    OCP Version  Documentation URL                                                                                                               
OpenShift removes several Kubernetes APIs in OpenShift 4.9.  false  4.9          https://access.redhat.com/documentation/en-us/openshift_container_platform/4.9/html/updating_clusters/updating-cluster-prepare  
```

What it looked like before the description length changed to be sorter

```
$ rosa list gates --profile jamesh2 --region us-east-1 --version 4.9.15                                                                   
Gate Description                                                        STS    OCP Version  Documentation URL                                                                                                                                 
OpenShift removes several Kubernetes APIs in OpenShift 4.9. To help     false  4.9          https://access.redhat.com/articles/6329921                                                                                                        
prevent issues with workloads and tools after upgrading, an                                                                                                                                                                                   
administrator is required to provide manual acknowledgement before the                                                                                                                                                                        
cluster can be upgraded from OpenShift 4.8 to 4.9.                                                                                                                                                                                            
                                                                                                                                                                                                                                              
OpenShift STS clusters include new required cloud provider permissions  true  4.9  https://docs.openshift.com/rosa/rosa_getting_started_sts/rosa_creating_a_cluster_with_sts/rosa-sts-about-iam-resources.html                                
in OpenShift 4.9. To prevent issues with the cluster upgrade, an                                                                                                                                                                              
administrator is required to provide manual acknowledgement that all                                                                                                                                                                          
STS roles are updated with 4.9 permissions before the cluster can be                                                                                                                                                                          
upgraded from OpenShift 4.8 to 4.9.
```

jharrington22 commented:
/retest

jharrington22 commented:
Fixed deps.

pvasant commented:
/lgtm
